ethereum/permissioning/src/test/java/tech/pegasys/pantheon/ethereum/permissioning/node/NodePermissioningControllerTest.java

```diff
@@ -0,0 +1,104 @@
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class NodePermissioningControllerTest {
+
+  private static final EnodeURL bootnode =
+      new EnodeURL(
+          "enode://6332792c4a00e3e4ee0926ed89e0d27ef985424d97b6a45bf0f23e51f0dcb5e66b875777506458aea7af6f9e4ffb69f43f3778ee73c81ed9d34c51c4b16b0b0f@192.168.0.1:9999");
+  private static final EnodeURL enode1 =
+      new EnodeURL(
+          "enode://94c15d1b9e2fe7ce56e458b9a3b672ef11894ddedd0c6f247e0f1d3487f52b66208fb4aeb8179fce6e3a749ea93ed147c37976d67af557508d199d9594c35f09@192.168.0.2:1234");
+  private static final EnodeURL enode2 =
+      new EnodeURL(
+          "enode://6f8a80d14311c39f35f516fa664deaaaa13e85b2f7493f37f6144d86991ec012937307647bd3b9a82abe2974e1407241d54947bbb39763a4cac9f77166ad92a0@192.168.0.3:5678");
+
+  @Mock private SyncStatusNodePermissioningProvider syncStatusNodePermissioningProvider;
+
+  private NodePermissioningController controller;
+  private Runnable hasReachedSyncCallback;
+
+  @Before
+  public void before() {
+    ArgumentCaptor<Runnable> captor = ArgumentCaptor.forClass(Runnable.class);
+    doNothing()
+        .when(syncStatusNodePermissioningProvider)
+        .setHasReachedSyncCallback(captor.capture());
+
+    this.controller = new NodePermissioningController(syncStatusNodePermissioningProvider);
+    this.hasReachedSyncCallback = captor.getValue();
+  }
+
+  @Test
+  public void isPermittedShouldDelegateToSyncStatusProvider() {
+    controller.isPermitted(enode1, enode2);
+
+    verify(syncStatusNodePermissioningProvider).isPermitted(eq(enode1), eq(enode2));
+  }
+
+  @Test
+  public void ifSyncHasBeenReachedPeerDiscoveryShouldRun() {
+    final Runnable peerDiscoveryCallback = mock(Runnable.class);
+    when(syncStatusNodePermissioningProvider.hasReachedSync()).thenReturn(true);
+
+    controller.startPeerDiscoveryCallback(peerDiscoveryCallback);
+
+    verify(peerDiscoveryCallback).run();
+  }
+
+  @Test
+  public void ifSyncHasNotBeenReachedShouldSetPeerDiscoveryCallback() {
+    final Runnable peerDiscoveryCallback = mock(Runnable.class);
+    when(syncStatusNodePermissioningProvider.hasReachedSync()).thenReturn(false);
+
+    controller.startPeerDiscoveryCallback(peerDiscoveryCallback);
+
+    hasReachedSyncCallback.run();
+
+    verifyZeroInteractions(peerDiscoveryCallback);
+    assertThat(controller.getPeerDiscoveryCallback()).isNotEmpty();
+  }
+
+  @Test
+  public void shouldExecutePeerDiscoveryCallbackWhenHasReachedSyncCallbackIsTriggered() {
+    final Runnable peerDiscoveryCallback = mock(Runnable.class);
+    when(syncStatusNodePermissioningProvider.hasReachedSync()).thenReturn(false);
+
+    controller.startPeerDiscoveryCallback(peerDiscoveryCallback);
+    hasReachedSyncCallback.run();
+
+    verify(peerDiscoveryCallback).run();
+    // should set the callback to empty after executing it once
```

Comment:
this test has the same setup as the previous one `ifSyncHasNotBeenReachedShouldSetPeerDiscoveryCallback` but the outcome is different?

Reply:
The outcome is different 'cause after running the callback we set it to empty. In the previous test we should not run the callback, just check if it has been set. I'll fix it! It is probably a CopyAndPasteException! ;)

Reply:
👍